src/inline-assembly.md

```diff
@@ -1,21 +1,34 @@
-# Inline assembly
+# Inline assembly [asm]
 
-Support for inline assembly is provided via the [`asm!`] and [`global_asm!`] macros.
-It can be used to embed handwritten assembly in the assembly output generated by the compiler.
+r[asm.macros]
+The macros [`core::arch::asm!`] and [`core::arch::global_asm!`] expand to inline assembly syntax when used in the expression position and item position respectively. The macros shall not be expanded in any other context.
 
-[`asm!`]: ../core/arch/macro.asm.html
-[`global_asm!`]: ../core/arch/macro.global_asm.html
+> [!NOTE]
+> The expansion of the macros has no stable syntax equivalent. This section will refer to the expansion of the macro, rather than the surface syntax.
 
-Support for inline assembly is stable on the following architectures:
+r[asm.safety]
+The macro [`core::arch::asm!`] shall be expanded only within an `unsafe` block.
```

Comment:
Does it? Shall is a static constraint on the program, so program is ill-formed if violated.

Reply:
As a casual reader, I agree that it could stand to be rephrased to mention a compiler error, or state some similarity to calling an `unsafe fn`.

Though the below compile_fail test, which you've asked to have stylistically called out as "this will fail to compile", might serve to alleviate the concern a little bit here. But it would be nice if this sentence could avoid being misleading in the first place.